packages/headless/src/features/placement-set/placement-set-slice.ts

```diff
@@ -0,0 +1,81 @@
+import {createReducer} from '@reduxjs/toolkit';
+import {
+  getBadge,
+  getRecs,
+  setCartSkus,
+  setLocale,
+  setOrderSkus,
+  setPlpSkus,
+  setProductSku,
+  setRecsSkus,
+  setSearchSkus,
+  setView,
+} from './placement-set-action';
+import {
+  getBadgesInitialState,
+  getPlacementSetInitialState,
+  getRecsInitialState,
+} from './placement-set-state';
+
+export const placementSetReducer = createReducer(
+  getPlacementSetInitialState(),
+  (builder) => {
+    builder
+      .addCase(setCartSkus, (state, action) => {
+        state.skus.cart = action.payload.skus;
+      })
+      .addCase(setOrderSkus, (state, action) => {
+        state.skus.order = action.payload.skus;
+      })
+      .addCase(setPlpSkus, (state, action) => {
+        state.skus.plp = action.payload.skus;
+      })
+      .addCase(setProductSku, (state, action) => {
+        state.skus.product = action.payload;
+      })
+      .addCase(setRecsSkus, (state, action) => {
+        state.skus.recs = action.payload.skus;
+      })
+      .addCase(setSearchSkus, (state, action) => {
+        state.skus.search = action.payload.skus;
+      })
+      .addCase(setLocale, (state, action) => {
+        state.view.currency = action.payload.currency;
+        state.view.locale = action.payload.locale;
+      })
+      .addCase(setView, (state, action) => {
+        state.view.type = action.payload.type ?? '';
+        state.view.subtype = action.payload.subtype;
+        state.view.type !== 'category' ? (state.skus.plp = []) : {};
```

Comment:
I understand that this action cleans up skus that are not to be used with this view type.

However, can we get in a problematic state if the `setSearchSkus` is called after the `setView` action with a `payload.type` of `product` , for example (search skus should never be used with the product view type)?

Reply:
You're correct, this approach could lead to weird states.

One thing I could do instead would be to have actions that would let the implementer set the correct SKUs at the same time as they set the rest of the current view context.

I'd keep the "setCartSkus" and "setRecsSkus" actions, because those two are relevant regardless of the current view context.

However, I'd replace the generic "setView" action and the other "set...Skus" actions with:

- usePdpContext(sku: string)
- usePlpContext(skus: string[], subtypes?: string[])
- useSearchContext(skus: string[])
- useCartContext()
- useCheckoutContext()
- useConfirmationContext()
- useHomeContext()

Each of those actions would set the proper view AND skus.

E.g., calling `usePlpContext(['123', '456'], ['category1', 'category2'])` would set state.view.type to 'category', state.view.subtypes to ['category1', 'category2'], and state.skus.plp to ['123', '456']. It would also automatically clean up any other useless skus in the state.

I think I like this approach better, because it means the implementer has fewer actions to dispatch, which makes it less error-prone.

This makes me think I could also add a "useCustomContext" action that would allow for wild stuff we haven't considered. However, that would imply that any string would be acceptable as a view "type" value... and I'm not convinced this generic custom action would actually be of any use, so I won't add it just now.

Reply:
I like this! Less stuff for an implementer to trip over :)